### ISSUE TITLE
can't map expression flattened by IncludeMembers()

### ISSUE THREAD

**Issue by samuelpsfung** [AutoMapper/AutoMapper.Extensions.ExpressionMapping#90]
In the following, I expect the expression is mapped to  `p => p.Stats.Speed`, but it throws exception "Cannot find member Speed of type Player".

```c#
    public class Player
    {
        public string Name { get; set; }
        public Stats Stats { get; set; }
    }

    public class Stats
    {
        public int Speed { get; set; }
        public int Power { get; set; }
    }

    public class PlayerModel
    {
        public string Name { get; set; }
        public int Speed { get; set; }
        public int Power { get; set; }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var mapper = new MapperConfiguration(c =>
            {
                c.AddExpressionMapping();
                c.CreateMap<Stats, PlayerModel>();
                c.CreateMap<Player, PlayerModel>()
                    .IncludeMembers(p => p.Stats);
            }).CreateMapper();

            Expression<Func<PlayerModel, int>> destExpr = m => m.Speed;
            var srcExpr = mapper.Map<Expression<Func<Player, int>>>(destExpr);
            Console.WriteLine(srcExpr.Body);
        }
    }
```

**Comment by BlaiseD** [comment]
I think you'll have to use `MapFrom`.  The following works.

```c#
     c.CreateMap<Player, PlayerModel>()
                    .ForMember(dest => dest.Speed, opts => opts.MapFrom(src => src.Stats.Speed));
```

I would have recommended a PR but I don't think a fix is possible in this library.  [propertyMap.SourceMembers](https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping/blob/eecf68c0167c32a3640dcbf6d5f22f5180d891b9/src/AutoMapper.Extensions.ExpressionMapping/XpressionMapperVisitor.cs#L491) only includes, `Speed` (instead of `Stats` and `Speed`).

Changing `PlayerModel.Speed` to `PlayerModel.StatsSpeed` works even without using `IncludeMembers`  (in which case `propertyMap.SourceMembers` includes both members).

Looks like you're seeing this problem because of the `IncludeMembers` implementation in the main [AutoMapper](https://github.com/AutoMapper/AutoMapper) library.  I would report it there.

In summary:  The property map for `Player.Stats.Speed` to `PlayerModel.Speed` only includes `Speed` in the source members list instead of `Stats` and `Speed`.

**Comment by samuelpsfung** [comment]
Hi Blaise, you can get `p => p.Stats` from

```c#
mapper.ConfigurationProvider
    .FindTypeMapFor<Player, PlayerModel>()
    .MemberMaps
    .Where(m => m.DestinationName == nameof(PlayerModel.Speed))
    .Single()
    .IncludedMember
    .MemberExpression
```

**Comment by BlaiseD** [comment]
@samuelpsfung - A PR is welcome.  Check out the code [here.](https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping/blob/master/src/AutoMapper.Extensions.ExpressionMapping/XpressionMapperVisitor.cs) and see it your recommendation works.

**Comment by samuelpsfung** [comment]
Hi Blaise,

I tried in my fork https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping/commit/38629e63beb85ac935cc05777db696316e01dfdd but I don't think it is ready for PR, can you please continue on that?

I've some questions
* the When_Use* testcases in ExpressionMapping don't seem calling Should_Validate()??
* I don't know how to make it work for UseAsDataSource()

**Comment by BlaiseD** [comment]
I think you're off track.  Adding another member to `PropertyMapInfo` isn't necessary.

You could do something like the following (making `GetMemberPath` public) -  also GetSourceMembers would be a local method in `FindDestinationFullName`:

```c#
            List<MemberInfo> GetSourceMembers(PropertyMap propertyMap, IMemberMap memberMap)
            {
                if (memberMap.CustomSource != null)
                {
                    return ExpressionHelpers.GetMemberPath
                    (
                        typeDestination, 
                        $"{memberMap.CustomSource.GetMemberFullName()}.{memberMap.DestinationName}"
                    ).ToList();
                }

                return propertyMap.SourceMembers.ToList();
            }
```

Them:

```c#
      propertyMapInfoList.Add(new PropertyMapInfo(propertyMap.CustomMapExpression, sourceMembers));
```

But even this looks a little messy.   I think the best fix is to head over to the AutoMapper repo and ask how you would add the parent reference - `Stats` in this case - to `propertyMap.SourceMembers` for `IncludeMembers`.

@lbargaoanu  can you advice please?

**Comment by BlaiseD** [comment]
Mapping for `UseAsDataSource` is implemented in a separate visitor [here](https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping/blob/eecf68c0167c32a3640dcbf6d5f22f5180d891b9/src/AutoMapper.Extensions.ExpressionMapping/ExpressionMapper.cs#L32).

**Comment by lbargaoanu** [comment]
You need to consider included members separately, because the implementation is different between `Map` and `ProjectTo`. You cannot rely on `SourceMembers` because that's shared.

**Comment by BlaiseD** [comment]
ok - thank you @lbargaoanu .

Sounds like you're saying `IncludeMembers` is used in `Map` but not `ProjectTo`. From the [flattening documentation](https://docs.automapper.org/en/stable/Flattening.html) I see it applies to `Map`. 

To be clear are you saying there is a downside in including the full source members list for include members?

So in `CreateMap`: `MapFrom`, `ForPath`, and "by convention" all use property maps, but `IncludeMembers` uses `IMemberMap`.

`ProjectTo` does not consider `IncludeMembers` but `Map` does?

For a given flattened member, both property map and member map exist on the `TypeMap` but depending on the configuration the parent reference may be represented as `IMemberMap.CustomSource` or as an item in `PropertyMap.SourceMembers`.

Sounds doable - don't know if it's worth the trouble.

**Comment by lbargaoanu** [comment]
No, as I've said in that other thread, `IncludeMembers` works with `ProjectTo`, and uses `ProjectToCustomSource`, not `SourceMembers`, so it doesn't interfere with `Map`. `IncludeMembers` works with `PropertyMap` too. `IMemberMap` is just an abstraction for all maps, `ForPath` works with `PathMap` and constructor mapping with `ConstructorMap`.

**Comment by BlaiseD** [comment]
ok - thanks

**Comment by BlaiseD** [comment]
That does make a little more sense if `ProperMap` has `CustomSource` as a member.  So this instead:
```c#
            List<MemberInfo> GetSourceMembers(PropertyMap propertyMap)
            {
                if (propertyMap.CustomSource != null)
                {
                    return ExpressionHelpers.GetMemberPath
                    (
                        typeDestination, 
                        $"{propertyMap.CustomSource.GetMemberFullName()}.{propertyMap.DestinationName}"
                    ).ToList();
                }

                return propertyMap.SourceMembers.ToList();
            }
```

Would have been messy logically if `CustomeSource` had to come from a whole different list of `MemberMap` .

**Comment by hasandogu** [comment]
I came across this thread while hoping to solve some problems related with the flattening outlined [here](https://github.com/AutoMapper/AutoMapper.Extensions.OData/pull/75#issuecomment-727288748), 

In order to achieve what I want, I needed to add support for IncludeMembers to this library.

@BlaiseD I have now implemented it as you outlined and also added support for UseAsDataSource in my fork here:

https://github.com/hasandogu/AutoMapper.Extensions.ExpressionMapping/commit/5d7ae2471971bca85049f0002a2c247581241bd1

These changes have fixed the unit tests I had added in the OData Extension as outlined in that [other thread](https://github.com/AutoMapper/AutoMapper.Extensions.OData/pull/75#issuecomment-727288748).

It is almost PR ready, except that it doesn't work when a custom map expression is involved in the data types. I added a unit test that fails because it cannot map the properties. I am not too familiar with the expression replacement, so I'd appreciate your help in getting this to the finish line.

The custom map expressions are handled here:

https://github.com/hasandogu/AutoMapper.Extensions.ExpressionMapping/blob/5d7ae2471971bca85049f0002a2c247581241bd1/src/AutoMapper.Extensions.ExpressionMapping/ExpressionMapper.cs#L214

How would you map these expressions to the source member properly?

**Comment by BlaiseD** [comment]
Flattening with `IncludeMembers()` should be working as of release [4.1.0-preview.0](https://www.nuget.org/packages/AutoMapper.Extensions.ExpressionMapping/4.1.0-preview.0).

**Comment by hasandogu** [comment]
Incredible! Thanks a lot, @BlaiseD.
